// @ark-ui/solid/dist/source/components/editable/use-editable.js
import * as editable from '@zag-js/editable';
import { normalizeProps, useMachine } from '@zag-js/solid';
import { createMemo, createUniqueId } from 'solid-js';
import { useEnvironmentContext, useLocaleContext } from '../../providers';
import { useFieldContext } from '../field';
export const useEditable = (props) => {
    const locale = useLocaleContext();
    const environment = useEnvironmentContext();
    const id = createUniqueId();
    const field = useFieldContext();
    const context = createMemo(() => ({
        id,
        ids: {
            label: field?.().ids.label,
            input: field?.().ids.control,
        },
        dir: locale().dir,
        disabled: field?.().disabled,
        invalid: field?.().invalid,
        readOnly: field?.().readOnly,
        required: field?.().required,
        getRootNode: environment().getRootNode,
        edit: props.defaultEdit,
        value: props.defaultValue,
        'edit.controlled': props.edit !== undefined,
        ...props,
    }));
    const [state, send] = useMachine(editable.machine(context()), { context });
    return createMemo(() => editable.connect(state, send, normalizeProps));
};
