// @ark-ui/solid/dist/source/components/tags-input/use-tags-input.js
import { normalizeProps, useMachine } from '@zag-js/solid';
import * as tagsInput from '@zag-js/tags-input';
import { createMemo, createUniqueId } from 'solid-js';
import { useEnvironmentContext, useLocaleContext } from '../../providers';
import { useFieldContext } from '../field';
export const useTagsInput = (props) => {
    const locale = useLocaleContext();
    const environment = useEnvironmentContext();
    const id = createUniqueId();
    const field = useFieldContext();
    const context = createMemo(() => ({
        id,
        ids: {
            label: field?.().ids.label,
            hiddenInput: field?.().ids.control,
        },
        dir: locale().dir,
        disabled: field?.().disabled,
        invalid: field?.().invalid,
        readOnly: field?.().readOnly,
        required: field?.().required,
        getRootNode: environment().getRootNode,
        value: props.defaultValue,
        ...props,
    }));
    const [state, send] = useMachine(tagsInput.machine(context()), { context });
    return createMemo(() => tagsInput.connect(state, send, normalizeProps));
};
