// @ark-ui/solid/dist/source/components/rating-group/use-rating-group.js
import * as rating from '@zag-js/rating-group';
import { normalizeProps, useMachine } from '@zag-js/solid';
import { createMemo, createUniqueId } from 'solid-js';
import { useEnvironmentContext, useLocaleContext } from '../../providers';
import { useFieldContext } from '../field';
export const useRatingGroup = (props) => {
    const locale = useLocaleContext();
    const environment = useEnvironmentContext();
    const id = createUniqueId();
    const field = useFieldContext();
    const context = createMemo(() => ({
        id,
        ids: {
            label: field?.().ids.label,
            hiddenInput: field?.().ids.control,
        },
        disabled: field?.().disabled,
        readOnly: field?.().readOnly,
        required: field?.().required,
        dir: locale().dir,
        getRootNode: environment().getRootNode,
        value: props.defaultValue,
        ...props,
    }));
    const [state, send] = useMachine(rating.machine(context()), { context });
    return createMemo(() => rating.connect(state, send, normalizeProps));
};
