// @ark-ui/solid/dist/source/components/signature-pad/use-signature-pad.js
import * as signaturePad from '@zag-js/signature-pad';
import { normalizeProps, useMachine } from '@zag-js/solid';
import { createMemo, createUniqueId } from 'solid-js';
import { useEnvironmentContext, useLocaleContext } from '../../providers';
import { useFieldContext } from '../field';
export const useSignaturePad = (props) => {
    const locale = useLocaleContext();
    const environment = useEnvironmentContext();
    const id = createUniqueId();
    const field = useFieldContext();
    const context = createMemo(() => ({
        id,
        ids: {
            label: field?.().ids.label,
            hiddenInput: field?.().ids.control,
        },
        disabled: field?.().disabled,
        readOnly: field?.().readOnly,
        required: field?.().required,
        dir: locale().dir,
        getRootNode: environment().getRootNode,
        ...props,
    }));
    const [state, send] = useMachine(signaturePad.machine(context()), { context });
    return createMemo(() => signaturePad.connect(state, send, normalizeProps));
};
