// @ark-ui/solid/dist/source/components/color-picker/use-color-picker.js
import * as colorPicker from '@zag-js/color-picker';
import { normalizeProps, useMachine } from '@zag-js/solid';
import { createMemo, createUniqueId } from 'solid-js';
import { useEnvironmentContext, useLocaleContext } from '../../providers';
import { useFieldContext } from '../field';
export const useColorPicker = (props) => {
    const locale = useLocaleContext();
    const environment = useEnvironmentContext();
    const field = useFieldContext();
    const id = createUniqueId();
    const context = createMemo(() => ({
        id,
        ids: {
            label: field?.().ids.label,
            input: field?.().ids.control,
        },
        dir: locale().dir,
        disabled: field?.().disabled,
        readOnly: field?.().readOnly,
        required: field?.().required,
        getRootNode: environment().getRootNode,
        open: props.defaultOpen,
        'open.controlled': props.open !== undefined,
        ...props,
        value: props.value
            ? colorPicker.parse(props.value)
            : props.defaultValue
                ? colorPicker.parse(props.defaultValue)
                : undefined,
    }));
    const [state, send] = useMachine(colorPicker.machine(context()), { context });
    return createMemo(() => colorPicker.connect(state, send, normalizeProps));
};
