// @ark-ui/solid/dist/source/components/file-upload/use-file-upload.js
import * as fileUpload from '@zag-js/file-upload';
import { normalizeProps, useMachine } from '@zag-js/solid';
import { createMemo, createUniqueId } from 'solid-js';
import { useEnvironmentContext, useLocaleContext } from '../../providers';
import { useFieldContext } from '../field';
export const useFileUpload = (props) => {
    const locale = useLocaleContext();
    const environment = useEnvironmentContext();
    const id = createUniqueId();
    const field = useFieldContext();
    const context = createMemo(() => ({
        id,
        ids: {
            label: field?.().ids.label,
            hiddenInput: field?.().ids.control,
        },
        dir: locale().dir,
        disabled: field?.().disabled,
        required: field?.().required,
        invalid: field?.().invalid,
        getRootNode: environment().getRootNode,
        ...props,
    }));
    const [state, send] = useMachine(fileUpload.machine(context()), { context });
    return createMemo(() => fileUpload.connect(state, send, normalizeProps));
};
