// @ark-ui/solid/dist/source/components/select/use-select.js
import * as select from '@zag-js/select';
import { normalizeProps, useMachine } from '@zag-js/solid';
import { createEffect, createMemo, createUniqueId, splitProps } from 'solid-js';
import { useEnvironmentContext, useLocaleContext } from '../../providers';
import { createSplitProps } from '../../utils/create-split-props';
import { useFieldContext } from '../field';
export const useSelect = (props) => {
    const [collectionOptions, selectProps] = createSplitProps()(props, [
        'isItemDisabled',
        'itemToValue',
        'itemToString',
        'items',
    ]);
    const collection = createMemo(() => select.collection({ ...collectionOptions }));
    const locale = useLocaleContext();
    const environment = useEnvironmentContext();
    const id = createUniqueId();
    const field = useFieldContext();
    const initialContext = createMemo(() => ({
        id,
        ids: {
            label: field?.().ids.label,
            hiddenSelect: field?.().ids.control,
        },
        disabled: field?.().disabled,
        readOnly: field?.().readOnly,
        invalid: field?.().invalid,
        required: field?.().required,
        collection: collection(),
        dir: locale().dir,
        getRootNode: environment().getRootNode,
        open: props.defaultOpen,
        value: props.defaultValue,
        'open.controlled': props.open !== undefined,
        ...selectProps,
    }));
    const context = createMemo(() => {
        const [, restProps] = splitProps(initialContext(), ['collection']);
        return restProps;
    });
    const [state, send] = useMachine(select.machine(initialContext()), {
        context,
    });
    const api = createMemo(() => select.connect(state, send, normalizeProps));
    createEffect(() => {
        api().setCollection(collection());
    });
    return api;
};
