// @ark-ui/solid/dist/source/components/checkbox/use-checkbox.js
import * as checkbox from '@zag-js/checkbox';
import { mergeProps, normalizeProps, useMachine } from '@zag-js/solid';
import { createMemo, createUniqueId } from 'solid-js';
import { useEnvironmentContext, useLocaleContext } from '../../providers';
import { useFieldContext } from '../field';
import { useCheckboxGroupContext } from './use-checkbox-group-context';
export const useCheckbox = (ownProps) => {
    const checkboxGroup = useCheckboxGroupContext();
    const props = createMemo(() => {
        return mergeProps(ownProps, checkboxGroup?.().getItemProps({ value: ownProps.value }) ?? {});
    }, [ownProps, checkboxGroup]);
    const locale = useLocaleContext();
    const environment = useEnvironmentContext();
    const id = createUniqueId();
    const field = useFieldContext();
    const context = createMemo(() => ({
        id,
        ids: {
            label: field?.().ids.label,
            hiddenInput: field?.().ids.control,
        },
        disabled: field?.().disabled,
        readOnly: field?.().readOnly,
        invalid: field?.().invalid,
        required: field?.().required,
        dir: locale().dir,
        getRootNode: environment().getRootNode,
        checked: props().defaultChecked,
        ...props(),
    }));
    const [state, send] = useMachine(checkbox.machine(context()), { context });
    return createMemo(() => checkbox.connect(state, send, normalizeProps));
};
