// @ark-ui/solid/dist/source/components/time-picker/use-time-picker.js
import { parseTime } from '@internationalized/date';
import { normalizeProps, useMachine } from '@zag-js/solid';
import * as timePicker from '@zag-js/time-picker';
import { createMemo, createUniqueId } from 'solid-js';
import { useEnvironmentContext, useLocaleContext } from '../../providers';
export const useTimePicker = (props = {}) => {
    const locale = useLocaleContext();
    const environment = useEnvironmentContext();
    const id = createUniqueId();
    const context = createMemo(() => ({
        id,
        dir: locale().dir,
        getRootNode: environment().getRootNode,
        open: props.defaultOpen,
        'open.controlled': props.open !== undefined,
        ...props,
        value: props.defaultValue ? parseTime(props.defaultValue) : undefined,
        min: props.min ? parseTime(props.min) : undefined,
        max: props.max ? parseTime(props.max) : undefined,
    }));
    const [state, send] = useMachine(timePicker.machine(context()), { context });
    return createMemo(() => timePicker.connect(state, send, normalizeProps));
};
